File: client/src/services/accountService.ts
import apiClient from "./apiClient";
import { ApiResponse } from "./authService";

export interface Account {
  id: string;
  tenantId: string;
  customerId?: string;
  accountNumber: string;
  accountType: string;
  currency: string;
  cachedBalance: number;
  status: string;
  createdAt: string;
  updatedAt: string;
}

export interface Customer {
  id: string;
  tenantId: string;
  name: string;
  externalReference?: string;
  email?: string;
  status: string;
  createdAt: string;
  accounts: Account[];
}

export interface CreateAccountRequest {
  accountNumber: string;
  accountType: string;
  currency: string;
  customerId?: string;
  initialDeposit?: number;
}

export interface CreateCustomerRequest {
  name: string;
  externalReference?: string;
  email?: string;
}

export const accountService = {
  getAccounts: async (customerId?: string): Promise<ApiResponse<Account[]>> => {
    try {
      const url = customerId ? `/accounts?customerId=${customerId}` : "/accounts";
      const response = await apiClient.get(url);
      return response.data;
    } catch (error: any) {
      return {
        success: false,
        errors: error.response?.data?.errors || ["Failed to fetch accounts"],
      };
    }
  },

  getAccountById: async (id: string): Promise<ApiResponse<Account>> => {
    try {
      const response = await apiClient.get(`/accounts/${id}`);
      return response.data;
    } catch (error: any) {
      return {
        success: false,
        errors: error.response?.data?.errors || ["Failed to fetch account"],
      };
    }
  },

  createAccount: async (request: CreateAccountRequest): Promise<ApiResponse<Account>> => {
    try {
      const response = await apiClient.post("/accounts", request);
      return response.data;
    } catch (error: any) {
      return {
        success: false,
        errors: error.response?.data?.errors || ["Failed to create account"],
      };
    }
  },

  getCustomers: async (): Promise<ApiResponse<Customer[]>> => {
    try {
      const response = await apiClient.get("/accounts/customers");
      return response.data;
    } catch (error: any) {
      return {
        success: false,
        errors: error.response?.data?.errors || ["Failed to fetch customers"],
      };
    }
  },

  createCustomer: async (request: CreateCustomerRequest): Promise<ApiResponse<Customer>> => {
    try {
      const response = await apiClient.post("/accounts/customers", request);
      return response.data;
    } catch (error: any) {
      return {
        success: false,
        errors: error.response?.data?.errors || ["Failed to create customer"],
      };
    }
  },
};

File: server/Services/AccountService.cs
using Tenvora.Api.Common;
using Tenvora.Api.Domain.Entities;
using Tenvora.Api.Dtos;
using Tenvora.Api.Repositories;

namespace Tenvora.Api.Services;

public class AccountService : IAccountService
{
    private readonly IAccountRepository _accountRepository;
    private readonly ICustomerRepository _customerRepository;

    public AccountService(IAccountRepository accountRepository, ICustomerRepository customerRepository)
    {
        _accountRepository = accountRepository;
        _customerRepository = customerRepository;
    }

    public async Task<AccountResponse?> GetAccountByIdAsync(Guid tenantId, Guid accountId)
    {
        var account = await _accountRepository.GetByIdAsync(tenantId, accountId);
        if (account == null) return null;

        return MapAccount(account);
    }

    public async Task<List<AccountResponse>> GetAccountsAsync(Guid tenantId, Guid? customerId = null)
    {
        var accounts = await _accountRepository.GetAllAsync(tenantId, customerId);
        return accounts.Select(MapAccount).ToList();
    }

    public async Task<AccountResponse> CreateAccountAsync(Guid tenantId, CreateAccountRequest request)
    {
        var existing = await _accountRepository.GetByAccountNumberAsync(tenantId, request.AccountNumber);
        if (existing != null)
        {
            throw new InvalidOperationException($"Account number {request.AccountNumber} already exists for this organization.");
        }

        var account = new Account
        {
            Id = Guid.NewGuid(),
            TenantId = tenantId,
            CustomerId = request.CustomerId,
            AccountNumber = request.AccountNumber.Trim(),
            AccountType = request.AccountType,
            Currency = request.Currency.ToUpperInvariant(),
            CachedBalance = request.InitialBalance,
            Status = AccountStatuses.Active,
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow
        };

        await _accountRepository.AddAsync(account);
        var created = await _accountRepository.GetByIdAsync(tenantId, account.Id);
        return MapAccount(created ?? account);
    }

    public async Task<CustomerResponse> CreateCustomerAsync(Guid tenantId, CreateCustomerRequest request)
    {
        var customer = new Customer
        {
            Id = Guid.NewGuid(),
            TenantId = tenantId,
            Name = request.Name.Trim(),
            Email = request.Email.Trim().ToLowerInvariant(),
            Phone = request.Phone,
            KycStatus = "Verified",
            Status = "Active",
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow
        };

        await _customerRepository.AddAsync(customer);
        return MapCustomer(customer);
    }

    public async Task<List<CustomerResponse>> GetCustomersAsync(Guid tenantId)
    {
        var customers = await _customerRepository.GetAllAsync(tenantId);
        return customers.Select(MapCustomer).ToList();
    }

    private static AccountResponse MapAccount(Account a) => new(
        a.Id,
        a.TenantId,
        a.CustomerId,
        a.Customer?.Name,
        a.AccountNumber,
        a.AccountType,
        a.Currency,
        a.CachedBalance,
        a.Status,
        a.CreatedAt,
        a.UpdatedAt
    );

    private static CustomerResponse MapCustomer(Customer c) => new(
        c.Id,
        c.TenantId,
        c.Name,
        c.Email,
        c.Phone,
        c.KycStatus,
        c.Status,
        c.Accounts.Select(MapAccount).ToList(),
        c.CreatedAt
    );
}
